src/services/signer_service.ts

```diff
@@ -0,0 +1,183 @@
+    }
+    public async generateExecuteTransactionEthereumTransactionAsync(
+        zeroExTransaction: ZeroExTransactionWithoutDomain,
+        signature: string,
+        protocolFee: BigNumber,
+    ): Promise<PartialTxParams> {
+        const gasPrice = zeroExTransaction.gasPrice;
+        const gas = await this._contractWrappers.exchange
+            .executeTransaction(zeroExTransaction, signature)
+            .estimateGasAsync({
+                from: META_TXN_RELAY_ADDRESS,
+                gasPrice,
+                value: protocolFee,
+            });
+
+        const executeTxnCalldata = await this._contractWrappers.exchange
+            .executeTransaction(zeroExTransaction, signature)
+            .getABIEncodedTransactionData();
+
+        const ethereumTxnParams: PartialTxParams = {
+            data: executeTxnCalldata,
+            gas: utils.encodeAmountAsHexString(gas),
+            from: META_TXN_RELAY_ADDRESS,
+            gasPrice: utils.encodeAmountAsHexString(gasPrice),
+            value: utils.encodeAmountAsHexString(protocolFee),
+            to: this._contractWrappers.exchange.address,
+            nonce: await this._getNonceAsync(META_TXN_RELAY_ADDRESS),
+            chainId: CHAIN_ID,
+        };
+
+        return ethereumTxnParams;
+    }
+    public async submitZeroExTransactionIfWhitelistedAsync(
+        zeroExTransaction: ZeroExTransactionWithoutDomain,
+        signature: string,
+        protocolFee: BigNumber,
+    ): Promise<PostTransactionResponse> {
+        const ethereumTxnParams = await this.generateExecuteTransactionEthereumTransactionAsync(
+            zeroExTransaction,
+            signature,
+            protocolFee,
+        );
+
+        const signedEthereumTransaction = await this._privateWalletSubprovider.signTransactionAsync(ethereumTxnParams);
+
+        const transactionHash = await this._contractWrappers.exchange
+            .executeTransaction(zeroExTransaction, signature)
+            .sendTransactionAsync({
+                from: META_TXN_RELAY_ADDRESS,
+                gasPrice: zeroExTransaction.gasPrice,
+                value: protocolFee,
+            });
+
+        return {
+            transactionHash,
+            signedEthereumTransaction,
+        };
+    }
+    private async _getNonceAsync(senderAddress: string): Promise<string> {
+        // HACK(fabio): NonceTrackerSubprovider doesn't expose the subsequent nonce
+        // to use so we fetch it from it's private instance variable
+        let nonce = (this._nonceTrackerSubprovider as any)._nonceCache[senderAddress];
+        if (nonce === undefined) {
+            nonce = await this._getTransactionCountAsync(senderAddress);
+        }
+        return nonce;
+    }
+    private async _getTransactionCountAsync(address: string): Promise<string> {
+        const nonceHex = await this._web3Wrapper.sendRawPayloadAsync<string>({
+            method: 'eth_getTransactionCount',
+            params: [address, 'latest'],
```

Comment:
`pending`?

Reply:
Yep, good catch. `pending` is def. safer here, since this code path will execute upon re-booting.